numba/compiler.py

```diff
@@ -163,19 +162,67 @@ def __call__(self):
             res = self.func()
         except _EarlyEscape as e:
             return _StageResult(self, ok=True, result=e.result,
-                               early_escape=True)
+                                early_escape=True)
         except BaseException as e:
             return _StageResult(self, ok=False, exception=e)
         else:
             return _StageResult(self, ok=True, result=res)
 
 
-def _raise_error(desc, exc):
-    """Patches the error
-    """
-    newmsg = "{desc}\n{exc}".format(desc=desc, exc=exc)
-    exc.args = (newmsg,)
-    return exc
+class _PipelineManager(object):
+    def __init__(self):
+        self.pipeline_order = []
+        self.pipeline_stages = {}
+        self._finalized = False
+
+    def create_pipeline(self, pipeline_name):
+        assert not self._finalized, "Pipelines can no longer be added"
+        self.pipeline_order.append(pipeline_name)
+        self.pipeline_stages[pipeline_name] = []
+        self.current = pipeline_name
+
+    def add_stage(self, stage_function, stage_description):
+        assert not self._finalized, "Stages can no longer be added."
+        new_stage = _Stage(stage_function, stage_description)
+        current_pipeline_name = self.pipeline_order[-1]
+        self.pipeline_stages[current_pipeline_name].append(new_stage)
+
+    def finalize(self):
+        self._finalized = True
+
+    def _patch_error(self, desc, exc):
+        """
+        Patches the error
+        """
+        newmsg = "{desc}\n{exc}".format(desc=desc, exc=exc)
+        exc.args = (newmsg,)
+        return exc
+
+    def run(self, status):
+        assert self._finalized, "PM must be finalized before run()"
+        res = None
+        for pipeline_name in self.pipeline_order:
+            for stage in self.pipeline_stages[pipeline_name]:
+                res = stage()
+                # Stage failed?
+                if not res.ok:
+                    # No more fallback pipelines?
+                    msg = "Failed at " + stage.msg
+                    is_final_pipeline = pipeline_name == self.pipeline_order[-1]
+                    if is_final_pipeline:
+                        raise self._patch_error(msg, res.exception)
+                    # Go to next fallback pipeline
+                    else:
+                        status.fail_reason = self._patch_error(msg, res.exception)
+                        break
+                # Stage OK and early escape
+                elif res.early_escape:
+                    return res.result
+            else:
+                return res.result
+
+        # TODO save all error information
+        raise CompilerError("All pipelines have failed")
```

Comment:
There is some pathological logic conditions on the _StageResult() object.  For example if rs.ok were True and res.early_escape were False, that could get you there, but based on perusing the _StageResult() initializers, these conditions should never arise.  

This suggests to me that there is some redundant signaling between ok and early_escape, but I don't understand well enough what they do to say for sure.  

What does the early_escap flag here really mean?